service/src/PlatformIO.cpp

```diff
@@ -296,26 +336,52 @@ namespace geopm
         if (ctl_tup_it != m_existing_control.end()) {
             result = ctl_tup_it->second;
         }
+        std::string err_msg;
+        bool read_signal_ok = false;
         if (result == -1) {
-            auto iogroup = find_control_iogroup(control_name);
-            if (iogroup != nullptr) {
-                if (iogroup->control_domain_type(control_name) == domain_type) {
-                    int group_control_idx = iogroup->push_control(control_name, domain_type, domain_idx);
-                    result = m_active_control.size();
-                    m_existing_control[ctl_tup] = result;
-                    m_active_control.emplace_back(iogroup, group_control_idx);
+            auto iogroups = find_control_iogroup(control_name);
+            for (auto ii : iogroups) {
+                if (ii->control_domain_type(control_name) == domain_type) {
+                    int val;
+                    try {
+                        // Attempt to read then write the control to ensure batch writes will succeed
+                        val = ii->read_signal(control_name, domain_type, domain_idx);
+                        ii->write_control(control_name, domain_type, domain_idx, val);
+                        read_signal_ok = true;
+                    }
+                    catch (const geopm::Exception &ex) {
+                        if (ex.err_value() == GEOPM_ERROR_NOT_IMPLEMENTED) {
+                            // IOGroups may not support read_signal()/write_control()
+                            read_signal_ok = true;
+                        }
+                        else {
+                            err_msg += std::string(ex.what()) + "\n";
+                        }
+                    }
+                    if (read_signal_ok == true) {
+                        int group_control_idx = ii->push_control(control_name, domain_type, domain_idx);
+                        result = m_active_control.size();
+                        m_existing_control[ctl_tup] = result;
+                        m_active_control.emplace_back(ii, group_control_idx);
+                        break;
```

Comment:
can we get rid of the early breaks and just put a

```
if (result != -1) {
    break;
}
```
At the end of these two loops over IOGroups?